### PR TITLE
Con 3301 view apprenticeship updates

### PR DESCRIPTION
**Pull request by hkf-tech**

### CODE REVIEW

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ConfirmEditApprenticeshipRequestToConfirmEditViewModelMapperTests.cs [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+
+            var result = await fixture.Map();
+
+            Assert.AreEqual(fixture.source.LastName, result.LastName);
+            Assert.AreEqual(fixture.source.CourseCode, result.CourseCode);
+        }
+
+
+        [Test]
+        public async Task NotChangedFieldsAreNull()
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ConfirmEditApprenticeshipRequestToConfirmEditViewModelMapperTests.cs:144]:
```suggestion
        public async Task UnchangedChangedFieldsAreNull()
```

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ReviewApprenticeshipUpdatesRequestToViewModelMapperTests.cs [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+        [Test]
+        public async Task TrainingName_IsMapped()
+        {
+            var viewModel = await fixture.Map();
+
+            Assert.AreEqual(fixture.ApprenticeshipUpdate.TrainingName, viewModel.ApprenticeshipUpdates.CourseName);
+        }
+
+
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ReviewApprenticeshipUpdatesRequestToViewModelMapperTests.cs:110]: double whitespace

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ReviewApprenticeshipUpdatesRequestToViewModelMapperTests.cs [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+                Mapper = new ReviewApprenticeshipUpdatesRequestToViewModelMapper(CommitmentApiClient.Object);
+            }
+
+            internal async Task<ReviewApprenticeshipUpdatesRequestViewModel> Map()
+            {
+                return await Mapper.Map(Source);
+            }
+        }
+
+
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ReviewApprenticeshipUpdatesRequestToViewModelMapperTests.cs:234]: double whitespace

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ReviewApprenticeshipUpdatesRequestToViewModelMapperTests.cs [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+                        dt = new DateTime(randomDateTime.Year, randomDateTime.Month, randomDateTime.Day);
+                    }
+                    else
+                    {
+                        dt = new DateTime(randomDateTime.Year, randomDateTime.Month, 1);
+                    }
+
+                    return dt;
+                }
+
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ReviewApprenticeshipUpdatesRequestToViewModelMapperTests.cs:259]: additional whitespace

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ReviewApprenticeshipUpdatesRequestToViewModelMapperTests.cs [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+            }
+        }
+
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ReviewApprenticeshipUpdatesRequestToViewModelMapperTests.cs:262]: additional whitespace

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ViewApprenticehipUpdatesRequestToViewModelMapperTests.cs [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+                Mapper = new ViewApprenticehipUpdatesRequestToViewModelMapper(CommitmentApiClient.Object);
+            }
+
+            internal async Task<ViewApprenticeshipUpdatesRequestViewModel> Map()
+            {
+                return await Mapper.Map(Source) as ViewApprenticeshipUpdatesRequestViewModel;
+            }
+        }
+
+
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ViewApprenticehipUpdatesRequestToViewModelMapperTests.cs:234]: whitespace

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ViewApprenticehipUpdatesRequestToViewModelMapperTests.cs [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+        public class DateTimeSpecimenBuilder : ISpecimenBuilder
+        {
+            public object Create(object request, ISpecimenContext context)
+            {
+                var pi = request as PropertyInfo;
+                if (pi == null || pi.PropertyType != typeof(DateTime?))
+                    return new NoSpecimen();
+
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ViewApprenticehipUpdatesRequestToViewModelMapperTests.cs:242]: could we include curly braces to this if statement as it is part of if/else

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/ApprenticeController.cs [hunk 8]
```diff
@@ -544,15 +547,105 @@ public async Task<IActionResult> EditApprenticeship(EditApprenticeshipRequestVie
...
+                var request = new UndoApprenticeshipUpdatesRequest
+                {
+                    ApprenticeshipId = viewModel.ApprenticeshipId,
+                    AccountId = viewModel.AccountId
+                };
+
+                await _commitmentsApiClient.UndoApprenticeshipUpdates(viewModel.ApprenticeshipId, request);
+            }
+
+            return await Task.FromResult(RedirectToAction(nameof(ApprenticeshipDetails), new { viewModel.AccountHashedId, viewModel.ApprenticeshipHashedId }));
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/ApprenticeController.cs:648]: no need for `Task.FromResult`

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/ApprenticeController.cs [hunk 9]
```diff
@@ -544,15 +547,105 @@ public async Task<IActionResult> EditApprenticeship(EditApprenticeshipRequestVie
...
+                var request = new RejectApprenticeshipUpdatesRequest
+                {
+                    ApprenticeshipId = viewModel.ApprenticeshipId,
+                    AccountId = viewModel.AccountId
+                };
+
+                await _commitmentsApiClient.RejectApprenticeshipUpdates(viewModel.ApprenticeshipId, request);
+            }
+
+            return RedirectToAction(nameof(ApprenticeshipDetails), new { viewModel.AccountHashedId, viewModel.ApprenticeshipHashedId });
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/ApprenticeController.cs:621]: no need for `Task.FromResult`

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ViewApprenticehipUpdatesRequestToViewModelMapper.cs [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+        {
+            _commitmentsApiClient = commitmentsApiClient;
+        }
+
+        public async Task<ViewApprenticeshipUpdatesRequestViewModel> Map(ViewApprenticehipUpdatesRequest source)
+        {
+            var updates = await _commitmentsApiClient.GetApprenticeshipUpdates(source.ApprenticeshipId,
+                   new CommitmentsV2.Api.Types.Requests.GetApprenticeshipUpdatesRequest { Status = CommitmentsV2.Types.ApprenticeshipUpdateStatus.Pending });
+
+            var apprenticeship = await _commitmentsApiClient.GetApprenticeship(source.ApprenticeshipId);
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ViewApprenticehipUpdatesRequestToViewModelMapper.cs:27]: Could we parallelize these calls?

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ViewApprenticehipUpdatesRequestToViewModelMapper.cs [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+            if (updates.ApprenticeshipUpdates.Count == 1)
+            {
+                var update = updates.ApprenticeshipUpdates.First();
+
+                var ApprenticeshipUpdates = GetApprenticeshipUpdates(update);
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ViewApprenticehipUpdatesRequestToViewModelMapper.cs:33]: lowecase for local vars

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ViewApprenticehipUpdatesRequestToViewModelMapper.cs [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+                LastName = apprenticeship.LastName,
+                DateOfBirth = apprenticeship.DateOfBirth,
+                ULN = apprenticeship.Uln,
+                StartDate = apprenticeship.StartDate,
+                EndDate = apprenticeship.EndDate,
+                CourseCode = apprenticeship.CourseCode
+            };
+
+            var courseDetailsOriginal = await _commitmentsApiClient.GetTrainingProgramme(apprenticeship.CourseCode);
+            OriginalApprenticeship.CourseName = courseDetailsOriginal?.TrainingProgramme.Name;
```
Comment (cofaulco) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ViewApprenticehipUpdatesRequestToViewModelMapper.cs:78]:
```suggestion
            OriginalApprenticeship.CourseName = apprenticeship.CourseName;
```

**Review by cofaulco — COMMENTED** [review]
a few comments

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ReviewApprenticeshipUpdatesRequestToViewModelMapperTests.cs (hunk 2 above)
Comment (hkf-tech) [reply src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ReviewApprenticeshipUpdatesRequestToViewModelMapperTests.cs:110]: done

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ReviewApprenticeshipUpdatesRequestToViewModelMapperTests.cs (hunk 3 above)
Comment (hkf-tech) [reply src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ReviewApprenticeshipUpdatesRequestToViewModelMapperTests.cs:234]: done

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ReviewApprenticeshipUpdatesRequestToViewModelMapperTests.cs (hunk 4 above)
Comment (hkf-tech) [reply src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ReviewApprenticeshipUpdatesRequestToViewModelMapperTests.cs:259]: done

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ReviewApprenticeshipUpdatesRequestToViewModelMapperTests.cs (hunk 5 above)
Comment (hkf-tech) [reply src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ReviewApprenticeshipUpdatesRequestToViewModelMapperTests.cs:262]: done

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ViewApprenticehipUpdatesRequestToViewModelMapperTests.cs (hunk 6 above)
Comment (hkf-tech) [reply src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ViewApprenticehipUpdatesRequestToViewModelMapperTests.cs:234]: done

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ViewApprenticehipUpdatesRequestToViewModelMapperTests.cs (hunk 7 above)
Comment (hkf-tech) [reply src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Mappers/Apprentice/ViewApprenticehipUpdatesRequestToViewModelMapperTests.cs:242]: done

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/ApprenticeController.cs (hunk 8 above)
Comment (hkf-tech) [reply src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/ApprenticeController.cs:707]: done

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ViewApprenticehipUpdatesRequestToViewModelMapper.cs (hunk 11 above)
Comment (hkf-tech) [reply src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Apprentice/ViewApprenticehipUpdatesRequestToViewModelMapper.cs:33]: done

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/ApprenticeController.cs (hunk 9 above)
Comment (hkf-tech) [reply src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/ApprenticeController.cs:680]: done

**Review by cofaulco — APPROVED** [review]

**Pull request merged**